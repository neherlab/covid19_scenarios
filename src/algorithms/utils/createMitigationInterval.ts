import _ from 'lodash'

import moment from 'moment'
import { v4 as uuidv4 } from 'uuid'
import createColor from 'create-color'

import { MitigationInterval, MitigationIntervals } from '../types/Param.types'

export function suggestNextMitigationInterval(intervals: MitigationIntervals): MitigationInterval {
  const tMaxMoments = intervals.map((interval) => moment(interval.timeRange.tMax))
  const tMinMoment = _.isEmpty(tMaxMoments) ? moment().startOf('day') : moment.max(tMaxMoments)
  const tMaxMoment = tMinMoment.clone().add(3, 'month')

  const timeRange = { tMin: tMinMoment.toDate(), tMax: tMaxMoment.toDate() }

  const interval: Omit<MitigationInterval, 'color'> = {
    id: uuidv4(),
    name: `Act from ${tMinMoment.format('D MMM YYYY')}`,
    timeRange,
    mitigationValue: 0.1,
  }

  const color = createColor(interval)

  return { ...interval, color }
}
