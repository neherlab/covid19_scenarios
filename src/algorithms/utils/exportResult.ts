import JSZip from 'jszip'
import { saveAs } from 'file-saver'
import { AlgorithmResult } from '../types/Result.types'
import { exportSimulation } from '../model'

export function isBlobApiSupported() {
  try {
    return !!new Blob()
  } catch (error) {
    return false
  }
}

export function saveFile(content: string, filename: string) {
  const blob = new Blob([content], { type: 'text/plain;charset=utf-8' })
  saveAs(blob, filename)
}

export async function exportAll(result: AlgorithmResult) {
  if (!result) {
    throw new Error(`Algorithm result expected, but got ${result}`)
  }

  const { deterministic, params } = result

  if (!(deterministic || params)) {
    console.error('Error: the results, and params, of the simulation cannot be exported')
    return
  }

  const zip = new JSZip()
  
  if (!params) {
    console.error('Error: the params of the simulation cannot be exported because they are null')
  } else {
    zip.file('covid.params.json', JSON.stringify(params, null, 2))
  }

  if (!deterministic) {
    console.error('Error: the results of the simulation cannot be exported because they are nondeterministic')
  } else {
    zip.file('covid.results.deterministic.csv', exportSimulation(deterministic))
  }

  const zipFile = await zip.generateAsync({ type: 'blob' })
  saveAs(zipFile, 'covid.params.results.zip')
}

export function exportResult(result: AlgorithmResult) {
  if (!result) {
    throw new Error(`Algorithm result expected, but got ${result}`)
  }

  const { deterministic } = result

  if (!isBlobApiSupported()) {
    // TODO: Display an error popup
    console.error('Error: export is not supported in this browser: `Blob()` API is not implemented')
    return
  }

  if (!deterministic) {
    console.error('Error: the results of the simulation cannot be exported because they are nondeterministic')
    return
  }

  saveFile(exportSimulation(deterministic), 'covid.results.deterministic.csv')
}

export function exportParams(result: AlgorithmResult) {
  if (!result) {
    throw new Error(`Algorithm result expected, but got ${result}`)
  }

  const { params } = result

  if (!isBlobApiSupported()) {
    // TODO: Display an error popup
    console.error('Error: export is not supported in this browser: `Blob()` API is not implemented')
    return
  }

  if (!params) {
    console.error('Error: the params of the simulation cannot be exported because they are null')
    return
  }

  saveFile(JSON.stringify(params, null, 2), 'covid.params.json')
}
