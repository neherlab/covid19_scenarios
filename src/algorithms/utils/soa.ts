import { map, zipObject } from 'lodash'

/**
 * Converts array of objects to an object of arrays ("Array of Structs" -> "Struct of Arrays" in olden terminology).
 * NOTE: The keys of the resulting object will be the same as in the *first* element of the input array.
 * NOTE: It will work properly with mismatched objects, but mostly only makes sense if all of the elements
 * of the input array are of the same shape and all properties are of the same type.
 */
export function soa<T extends { [key: string]: unknown }, K extends keyof T>(arr: T[]): { [key: string]: T[K][] } {
  if (arr.length === 0) {
    return {} as { [key: string]: T[K][] }
  }

  const keys = Object.keys(arr[0])
  const something = keys.map((key) => map(arr, key))
  return zipObject(keys, something) as { [key: string]: T[K][] }
}
