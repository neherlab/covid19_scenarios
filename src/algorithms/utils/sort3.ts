/* eslint-disable no-param-reassign */

/** Puts 3 given numbers in ascending order. Does not mutate arguments. */
export function sort3<T extends number>(a: T, b: T, c: T) {
  if (a > c) [a, c] = [c, a]
  if (a > b) [a, b] = [b, a]
  if (b > c) [b, c] = [c, b]
  return [a, b, c]
}
