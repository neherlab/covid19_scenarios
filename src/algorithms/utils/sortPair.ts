/** Puts 2 given numbers in ascending order. Does not mutate arguments. */
export function sortPair<T extends number>([a, b]: [T, T]) {
  if (a < b) return [a, b]
  return [b, a]
}
