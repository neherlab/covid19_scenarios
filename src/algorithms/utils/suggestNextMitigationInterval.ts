import moment from 'moment'

import createColor from 'create-color'
import type { StrictOmit } from 'ts-essentials'
import { uuidv4 } from '../../helpers/uuid'

import type { MitigationInterval } from '../types/Param.types'

export function suggestNextMitigationInterval(): MitigationInterval {
  const tMinMoment = moment()
  const tMaxMoment = tMinMoment.clone().add(30, 'days')

  const timeRange = { begin: moment().toDate(), end: tMaxMoment.toDate() }

  const interval: StrictOmit<MitigationInterval, 'color'> = {
    id: uuidv4(),
    name: `Intervention from ${tMinMoment.format('D MMM YYYY')}`,
    timeRange,
    transmissionReduction: { begin: 10, end: 30 },
    mitigationAgeSpecificData: [],
  }

  const color = createColor(interval)

  return { ...interval, color }
}
