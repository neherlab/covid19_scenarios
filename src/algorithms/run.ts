import * as math from 'mathjs'

import { OneCountryAgeDistribution } from '../assets/data/CountryAgeDistribution.types'

import { SeverityTableRow } from '../components/Main/Scenario/SeverityTable'

import { collectTotals, evolve, getPopulationParams, initializePopulation } from './model'
import { AllParamsFlat, MitigationIntervals } from './types/Param.types'
import { AlgorithmResult, SimulationTimePoint } from './types/Result.types'
import { TimeSeries } from './types/TimeSeries.types'

const identity = (x: number) => x
const poisson = (x: number) => {
  throw new Error('We removed dependency on `random` package. Currently `poisson` is not implemented')
}

interface ChangePoint {
  t: Date
  val: Number[]
}

const compareTimes = (a: ChangePoint, b: ChangePoint): number => {
  return a.t.valueOf() - b.t.valueOf()
}

// NOTE: Assumes containment is sorted ascending in time.
export function interpolateTimeSeries(containment: TimeSeries): (t: Date) => number {
  if (containment.length === 0) {
    return (t: Date) => 1.0
  }
  const Ys = containment.map((d) => d.y)
  const Ts = containment.map((d) => Number(d.t))
  return (t: Date) => {
    if (t <= containment[0].t) {
      return containment[0].y
    }
    if (t >= containment[containment.length - 1].t) {
      return containment[containment.length - 1].y
    }
    const i = containment.findIndex((d) => Number(t) < Number(d.t))

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const evalLinear = (t: number) => {
      const deltaY = Ys[i] - Ys[i - 1]
      const deltaT = Ts[i] - Ts[i - 1]

      const dS = deltaY / deltaT
      const dT = t - Ts[i - 1]

      return Ys[i - 1] + dS * dT
    }

    return evalLinear(Number(t))
  }
}


export function intervalsToTimeSeries(intervals: MitigationIntervals): TimeSeries {
  const changePoints = {}
  intervals.forEach((element) => {
    // bound the value by 0.01 and 100 (transmission can be at most 100 fold reduced or increased)
    const val = Math.min(Math.max(1 - element.mitigationValue, 0.01), 100)

    if (changePoints[element.timeRange.tMin] !== undefined){
      changePoints[element.timeRange.tMin].push(val)
    } else {
      changePoints[element.timeRange.tMin] = [val]
    }
    // add inverse of the value when measure is relaxed
    if (changePoints[element.timeRange.tMax] !== undefined){
      changePoints[element.timeRange.tMax].push(1.0/val)
    } else {
      changePoints[element.timeRange.tMax] = [1.0/val]
    }
  });

  const orderedChangePoints: ChangePoint[] = Object.keys(changePoints).map(d => ({t:new Date(d), val:changePoints[d]}))

  orderedChangePoints.sort(compareTimes)

  if (orderedChangePoints.length) {
    const mitigationSeries: TimeSeries = [{t:orderedChangePoints[0].t, y:1.0}]
    const product = (a: Number,b: Number): Number => (a*b);

    orderedChangePoints.forEach((d,i) => {
      const prevValue = mitigationSeries[2*i].y
      const newValue = d.val.reduce(product, prevValue)
      mitigationSeries.push({t:d.t, y:prevValue})
      mitigationSeries.push({t:d.t, y:newValue})
    })
    return mitigationSeries
  }
  return [];
}

/**
 *
 * Entry point for the algorithm
 *
 */
export async function run(
  params: AllParamsFlat,
  severity: SeverityTableRow[],
  ageDistribution: OneCountryAgeDistribution,
  containment: TimeSeries,
): Promise<AlgorithmResult> {
  const modelParams = getPopulationParams(params, severity, ageDistribution, interpolateTimeSeries(containment))
  const tMin: number = new Date(params.simulationTimeRange.tMin).getTime()
  const tMax: number = new Date(params.simulationTimeRange.tMax).getTime()
  const initialCases = params.suspectedCasesToday
  let initialState = initializePopulation(modelParams.populationServed, initialCases, tMin, ageDistribution)

  function simulate(initialState: SimulationTimePoint, func: (x: number) => number) {
    const dynamics = [initialState]
    let currState = initialState
    while (currState.time < tMax) {
      currState = evolve(currState, modelParams, currState.time + 1, func)
      dynamics.push(currState)
    }

    return collectTotals(dynamics)
  }

  const sim: AlgorithmResult = {
    deterministic: simulate(initialState, identity),
    stochastic: [],
    params: modelParams,
  }

  for (let i = 0; i < modelParams.numberStochasticRuns; i++) {
    initialState = initializePopulation(modelParams.populationServed, initialCases, tMin, ageDistribution)
    sim.stochastic.push(simulate(initialState, poisson))
  }

  return sim
}
