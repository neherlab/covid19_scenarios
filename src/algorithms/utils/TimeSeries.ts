import * as d3 from 'd3'

import { DateRange } from '../types/Param.types'
import { interpolateTimeSeries } from '../run'
import { TimeSeries } from '../types/TimeSeries.types'

export function uniformDatesBetween(min: number, max: number, n: number): Date[] {
  const d = (max - min) / (n - 1)
  const dates = d3.range(min, max + d, d).filter((_, i) => i < n)
  return dates.map((d) => new Date(d))
}

export function makeTimeSeries(simulationTimeRange: DateRange, values: number[]): TimeSeries {
  const { tMin, tMax } = simulationTimeRange
  const n = values.length

  const dates = uniformDatesBetween(tMin.getTime(), tMax.getTime(), n)

  const tSeries = []
  for (let i = 0; i < n; i++) {
    tSeries.push({ t: dates[i], y: values[i] })
  }

  return tSeries
}

export function updateTimeSeries(simulationTimeRange: DateRange, oldTimeSeries: TimeSeries, n: number): TimeSeries {
  const { tMin, tMax } = simulationTimeRange
  const interpolator = interpolateTimeSeries(oldTimeSeries)
  const clamp = function(x: number, min: number, max: number): number {
      if (x < min) {
          return min;
      } else if (x > max) {
          return max;
      } else {
          return x;
      }
  }


  const dates = uniformDatesBetween(tMin.getTime(), tMax.getTime(), n) 
  return dates.map((d) => ({ t: d, y: clamp(interpolator(d), 0, 1.2) }));
}
