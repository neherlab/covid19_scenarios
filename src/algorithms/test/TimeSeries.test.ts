import { uniformDatesBetween, makeTimeSeries, updateTimeSeries } from '../utils/TimeSeries'
import { interpolateTimeSeries } from '../run'

function compareTimeSeries(expected: { t: Date; y: number }[], result: { t: Date; y: number }[]) {
  expect(result).toHaveLength(expected.length)

  expected.forEach(({ t, y }, index) => {
    const { t: resultT, y: resultY } = result[index]
    expect(resultT).toStrictEqual(t)
    expect(resultY).toBeCloseTo(y)
  })
}

describe('TimeSeries', () => {
  const tMin = new Date('1970-10-01T00:00:00.000Z')
  const tMax = new Date('1971-02-01T00:00:00.000Z')

  // These are the expected values from uniformDatesBetween(tMin,tMax)
  const tBetween = [
    new Date('1970-10-31T18:00:00.000Z'),
    new Date('1970-12-01T12:00:00.000Z'),
    new Date('1971-01-01T06:00:00.000Z'),
  ]

  describe('uniformDatesBetween()', () => {
    it('returns an array of "n" uniformly spaced dates between "min" and "max", inclusive.', () => {
      const n = 5

      const result = uniformDatesBetween(tMin.getTime(), tMax.getTime(), n)

      expect(result).toBeArray()

      expect(result).toStrictEqual([tMin, ...tBetween, tMax])
    })
  })

  describe('makeTimeSeries()', () => {
    it('maps an array of length "n" values to an array of "n" uniform dates bracketed by the simulation time range.', () => {
      const simulationTimeRange = { tMin, tMax }

      const result = makeTimeSeries(simulationTimeRange, [1, 2, 3, 4, 5])

      expect(result).toBeArray()

      expect(result).toStrictEqual([
        { t: tMin, y: 1 },
        { t: tBetween[0], y: 2 },
        { t: tBetween[1], y: 3 },
        { t: tBetween[2], y: 4 },
        { t: tMax, y: 5 },
      ])
    })
  })

  describe('interpolateTimeSeries()', () => {
    it('if created using an empty vector, the interpolation function must throw', () => {
      expect(() => interpolateTimeSeries([])).toThrow()
    })

    it('if given a time less than tMin, the interpolation function returns the earliest timeseries "y" value.', () => {
      const simulationTimeRange = { tMin, tMax }
      const timeSeries = makeTimeSeries(simulationTimeRange, [1, 3, 2, 5, 4, 7])

      const result = interpolateTimeSeries(timeSeries)

      expect(result).toBeFunction()

      const earlier = new Date(tMin.getTime() - 1000)
      expect(result(earlier)).toBe(1)
    })

    it('if given a time greater than tMax, the interpolation function returns the oldest timeseries "y" value.', () => {
      const simulationTimeRange = { tMin, tMax }
      const timeSeries = makeTimeSeries(simulationTimeRange, [1, 3, 2, 5, 4, 7])

      const result = interpolateTimeSeries(timeSeries)

      expect(result).toBeFunction()

      const later = new Date(tMax.getTime() + 1000)
      expect(result(later)).toBe(7)
    })

    it('interpolation within the timerange occurs as expected.', () => {
      const simulationTimeRange = { tMin, tMax }
      const timeSeries = makeTimeSeries(simulationTimeRange, [1, 7, 10])

      const interpolator = interpolateTimeSeries(timeSeries)

      expect(interpolator).toBeFunction()

      /* Run the interpolator 10 times the [tMin,tMax] range, stepping
       * in intervals of 15 days */

      const intervalCount = 10
      const intervalSize = 15

      const result = [...new Array(intervalCount)].map((_, index) => {
        const date = new Date(Number(tMin))
        date.setDate(date.getDate() + intervalSize * index)
        return interpolator(date)
      })

      expect(result).toBeArrayOfSize(intervalCount)

      const expected = [
        1,
        2.6354594390679185,
        4.209836664611696,
        5.648965995858905,
        6.892037521792783,
        7.889772449043392,
        8.678773876200733,
        9.32431394905387,
        9.891684813683101,
        10,
      ]

      result.forEach((y, index) => {
        expect(y).toBeCloseTo(expected[index])
      })
    })
  })

  describe('updateTimeSeries()', () => {
    it('interpolates an existing TimeSeries with a new "y" vector.', () => {
      const simulationTimeRange = { tMin, tMax }
      const yVector = [3, 7, 11]
      const timeSeries = makeTimeSeries(simulationTimeRange, yVector)
      const n = 5

      /* Update the time series using the new 'n' value with same TimeRange. */
      const result = updateTimeSeries(simulationTimeRange, timeSeries, n)
      expect(result).toBeArrayOfSize(n).toMatchSnapshot()
    })

    it('interpolates an existing TimeSeries with a new TimeRange.', () => {
      const simulationTimeRange = { tMin, tMax }
      const yVector = [3, 7, 11]
      const timeSeries = makeTimeSeries(simulationTimeRange, yVector)

      /* Add/remove 5 days from tMin and tMax, respectively,
       * resulting in a TimeRange smaller by 10 days.
       */
      const newTMin = new Date(Number(tMin))
      newTMin.setDate(newTMin.getDate() + 5)
      const newTMax = new Date(Number(tMax))
      newTMax.setDate(newTMax.getDate() - 5)
      const newSimulationTimeRange = { tMin: newTMin, tMax: newTMax }

      /* Update the time series using the new TimeRange with same 'n' value */
      const result = updateTimeSeries(newSimulationTimeRange, timeSeries, yVector.length)

      expect(result).toBeArrayOfSize(yVector.length).toMatchSnapshot()
    })
  })
})
