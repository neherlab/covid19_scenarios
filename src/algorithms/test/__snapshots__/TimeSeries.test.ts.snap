// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeSeries updateTimeSeries() interpolates an existing TimeSeries with a new "y" vector. 1`] = `
Array [
  Object {
    "t": 1970-10-01T00:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1970-10-31T18:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1970-12-01T12:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1971-01-01T06:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1971-02-01T00:00:00.000Z,
    "y": 1.2,
  },
]
`;

exports[`TimeSeries updateTimeSeries() interpolates an existing TimeSeries with a new TimeRange. 1`] = `
Array [
  Object {
    "t": 1970-10-06T00:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1970-12-01T12:00:00.000Z,
    "y": 1.2,
  },
  Object {
    "t": 1971-01-27T00:00:00.000Z,
    "y": 1.2,
  },
]
`;
